### DISCUSSION TITLE
Welcome to sys-con Discussions!

### DISCUSSION THREAD

**Discussion by o0Zz** [o0Zz/sys-con#11] [Announcements]
## 👋 Welcome !
  This place is open to any user that want to : 
  * Share ideas.
  * Share a new configuration for a controller not yet supported
  * Ask for questions

**Comment by texita** [comment]
One suggestion about having the logs timestamp using the switch time instead of the switch boot time. (or maybe both in debug/trace?)

I was thinking on when going to meet some friends, they have got joypads different from mines and at them house i'm without pc and don't have time to do nerding stuff, but i can connect a controller for some seconds and even if it's not working i will have the correct PID/VID, i'll do a screenshot on the switch or a picture with the phone, and that's it, done xD
PID/VID stealing FTW!

**Reply by o0Zz** [reply]
Using system time is not really reliable for debugging and doesn't give millisecond accuracy. If I understand your request correctly, you just want to get your friend's vid/pid. How about a new feature that automatically adds a vid/pid to the config.ini if it's not there. (A kind of "auto_add_controllers=1") and when a controller is plugged in and unknown, we add the vid/pid and a "hardcoded" configuration.
This way you can read your vid/pid at home and maybe have a halfway "working" controller where you can see which button is not in the right place and fix it at home. What do you think? Does it suit your needs?

**Reply by texita** [reply]
sounds really good! No rush of course, was only an idea, but your is better :)

**Reply by o0Zz** [reply]
FYI the latest sys-con release add this auto add controller feature

**Comment by texita** [comment]
want to share my config for this pro controller (sorry for wrong discussion post, i saw later the right discussion where i should post this)

[0e6f-0184] ; PDP gaming faceoff deluxe+ audio wired controller
b=2
a=3
x=4
y=1
l=5
r=6
zl=7
zr=8
minus=9
plus=10
capture=14
home=13
lstick_click=11
rstick_click=12

**Comment by texita** [comment]
i was thinkin another thing: what about having all the default profiles buttons binding to match official switch buttons layout? i explain what i mean, i've tried lot of original pads (ps3,ps4,ps5,xboxone and more will come) i see that some default mappings (like dualshock3) are following the default switch layout, example triangle is X and circle is A. Meanwhile xboxone for example is following the correct xbox pad layout where Y on the pad is Y on the switch, but is physically inverted on the pad layout.
At the end, if you are using a pad  on the switch i think is better to have buttons to match by default the switch layout indipedent from the pad that you are using, plus the fact is more coherent when you are trying buttons :)

**Reply by o0Zz** [reply]
When you say _**I see that some default mappings (like dualshock3) follow the default switch layout**_ you mean that they match the id set in the configuration (1, 2, 3 etc...) ? 

What you need to understand here is that there is no way to determine where the triangle is. Generic controllers exposes a "profile" with a list of buttons (With an ID), but we don't know if 1 is a triangle or not. This means that button 1 could be a triangle on one controller and a circle on another.
If you feel that there is a pattern that comes up every time, then yes, we can try to have a default mapping that matches most of them. However, I haven't found that kind of combination on my side (I don't have hundreds of controllers ;) - just 5 or 6.)

For now the hardcoded values are :
```
b=1
a=2
x=3
y=4
l=5
r=6
zl=7
zr=8
Minus=9
plus=10
catch=11
home=12
lstick_click=13
rstick_click=14
```
I think we can change it to whatever we want (I have no problem with that). 

However, it's not a good idea to change the `[default]` mapping (not sure if that's what you wanted to do originally). The reason is, everything is set to 0 for one reason, it means "Disabled", so by default all buttons are disabled (Unmapped), if we decide to change that, then you will have to override 100% of the buttons in all other profiles (and not just the one you want to use) That the counter part which will quickly become annoying ;)

**Comment by texita** [comment]
Nono i'm sorry, i'm expressed really bad, i was talking about profiles section and that types of default bindings. i don't wanna touch the [default] 🙈

Going back to what i was talking about, i will use this in ps5 profile, so if you press triangle in a original ps5 joypad is like pressing X, because X on the switch controllers is on the top, same for A (circle) and B (x)

[dualsense5] ;PS5
color_body = #f7f7f7
color_buttons = #c6c6c6
color_leftGrip = #474747
color_rightGrip = #474747
B=2
A=3
Y=1
X=4
L=5
R=6
ZL=7
ZR=8
minus=9
plus=10
home=13
capture=14
lstick_click=11
rstick_click=12

Same example for the xboxone 
[xboxone]
driver=xboxone
color_body = #f1f1f1
color_buttons = #020a0c
color_leftGrip = #b4b4b4
color_rightGrip = #b4b4b4
B=1
A=2
Y=3
X=4
L=8
R=9
ZL=0
ZR=0
minus=7
plus=6
home=12
capture=5
lstick_click=10
rstick_click=11

this is the dualshock3 part, copied from your config, is already matching the switch layout so i didn't change anything
[dualshock3]
driver=dualshock3
color_body = #1c1c1c
color_buttons = #515050
color_leftGrip = #1c1c1c
color_rightGrip = #1c1c1c
B=3
A=2
Y=4
X=1
L=7
R=8
ZL=5
ZR=6
minus=9
plus=12
lstick_click=10
rstick_click=11
home=13

this is xbox360 in this way, for example i removed buttons from your config and my horipad leaving only profile=xbox360

[0f0d-00dc] ;HORIPAD FPS for Nintendo Switch
profile=xbox360

then i changed the profile section

[xbox360]
driver=xbox360
color_body = #f1f1f1
color_buttons = #810f0f
color_leftGrip = #b4b4b4
color_rightGrip = #b4b4b4
B=2
A=1
Y=4
X=3
L=5
R=6
ZL=0
ZR=0
minus=8
plus=7
home=11
lstick_click=9
rstick_click=10

right now i can't test with an original xbox360 controller, but i think if you try with your x360 pads, they should match the switch layout with this config, right?

i've tested:
ps3 with orginals dualshock3 and sixaxis controllers
ps5 with dualsense (not edge one, but should be the same)
xbox one with original controllers

i can't test right now cause i left controllers to friends: ps4,xbox360
can't test because i haven't got adapters or pads: xbox,xbox360w

**Reply by o0Zz** [reply]
Ok I better understand your point, so yes this is indeed exactly how is suppose to work profiles, they are supposed to match the switch button layout.

Few comments:

For dualsense5, this mapping comes from another user (I don't have dualsense5) thus I think he didn't understand how it was suppose to work. I will update it with yours.

For xboxone, it's interresting, I don't have a xboxone neither but the mapping I used for this one is supposed to be the same than the old sys-con but I maybe did a misstake while recreating it. Could you let me know if you use this mapping with an official xboxone ?

For xbox360, this one is correct and match what you said for most of controllers. Indeed it has been created from original xbox360 controller and you are right the ";HORIPAD FPS for Nintendo Switch" is an exception and do not match the default layout that's why there a dedicated section for this one:

```
profile=xbox360
B=2
A=1
Y=4
X=3
```

So, to me xbox360 is correct, dualsense5 I will updated it, and I'm waiting for feedback for the xboxone one.
Thanks

**Comment by texita** [comment]
xboxone is matching joypad layout not switch, so need to be corrected. I've tested with wireless controller for xbox one model 1537 [045e-02d1]

also ps5 as you said need to be corrected because is following the xbox layout, not switch. tested with DualSense [054c-0ce6]

i'm gonna test ps4 tonight probably

**Reply by o0Zz** [reply]
Just to be sure i don't do another misstakew ;), could you send me the final xboxone and PS5 mapping please ?

**Reply by texita** [reply]
sure, need only to invert A with B and X with Y so:

[xboxone]
driver=xboxone
B=1
A=2
Y=3
X=4
L=8
R=9
ZL=0
ZR=0
minus=7
plus=6
home=12
capture=5
lstick_click=10
rstick_click=11

[dualsense5] ;PS5
B=2
A=3
Y=1
X=4
L=5
R=6
ZL=7
ZR=8
minus=9
plus=10
home=13
capture=14
lstick_click=11
rstick_click=12

Tested dualshock 4 with [054c-09cc] DS4 v2 and everything ok, matching switch layout

**Comment by kfkboys** [comment]
I'm having a hard time trying to do something odd.

I got Ace Combat 7 and I have the switch setup as normal however I want to use a HOTAS setup.
I've got the PS3 HOTAS and several other HOTAS setups but I've got almost everything working properly on the PS3 version of the setup - Truthfully I want to figure this out so I can get the PS4 version setup so I can attach the pedals to hook into the YAW controls and get a real flight sim setup on the switch. 
Everything currently works except the camera setup is odd (it does work but I want to setup to use the paddle - Left and Right - for YAW) and throttle is set to buttons not to the left stick because I have yet to figure out how to map individual stick directions to a button.

Ultimately I want to map or swap the following:
left stick LEFT (Currently Throttle Paddle) to ZL
left stick RIGHT (Currently Throttle Paddle) to ZR
left stick UP (Currently Throttle) to R
left stick DOWN (Currently Throttle) to L

and then I can then map or swap the current buttons to the camera currently left stick
ZR (Currently Button 8) to something else probably left stick to right direction
ZL (Currently Button 7) to something else probably left stick to left direction
R (Currently Button 1) to something else probably left stick up direction
L (Currently Button 4) to something else probably left stick down direction

I would also like if I can map things to the D-PAD, but ideally still map things to each stick direction would be amazing.

I see the swapDPADandLSTICK in the SwitchAbstractedPadHandler.cpp and it's possible to swap these L-stick to dpad but I am failing to see how I can do this for individual buttons. I'm sure I'm missing something simple.
One of the only HOTAS setups I see in here is the X52FlightControlSystem and I unfortunately do not own that one.

Here's my setup thus far:

[044f-b109] ;Automatically added on 2024-7-13 0:30:53UTC
b=6
a=5
x=11
y=12
l=4
r=1
zl=8
zr=7
minus=9
plus=10
capture=15
home=13
lstick_click=2
rstick_click=3

**Reply by o0Zz** [reply]
I think I understand the request and you're right today it's not possible to assign directions on buttons and buttons on directions but this is something we can probably add as a feature. Could you please create a ticket "feature request" and post this message, we will discuss about this feature in this ticket. Thanks
